## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.columns

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.


In [5]:
# What three item categories had the most entries?
top_categories = df['category'].value_counts().head(3)
top_categories

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries
most_common_cat = top_categories.index[0]
top_sub_cat = df[df['category'] == most_common_cat]['subcategory'].value_counts().head(1)
top_sub_cat

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
top5_clients = df['client_id'].value_counts().head(5)
top5_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top5_clients_list = top5_clients.index.tolist()
top5_clients_list


[33615, 66037, 46820, 38378, 24741]

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
client_with_most_entries = top5_clients.index[0]
total_units_ordered = df[df['client_id'] == client_with_most_entries]['qty'].sum()
total_units_ordered


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_price'] = df['unit_weight'].apply(lambda x: 7 * x if x > 50 else 10 * x)
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0            75.0  
1            14.9  
2            16.8  
3            12.3  
4           464.3

In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00            75.0   
1      15.09         1.49   21            0     523.95            14.9   
2       7.86         1.68   39            6     527.28            16.8   
3      24.85         1.23   29            3    1056.18            12.3   
4     108.17        46.43   20            1    3902.00           464.3   

     total_price  
0  125898.607500  
1     588.693625  
2     594.407400  
3    1167.314400  
4    4770.182750  

[5 rows x 21 columns]

In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00            75.0  125898.607500   
1         1.49   21            0     523.95            14.9     588.693625   
2         1.68   39            6     527.28            16.8     594.407400   
3         1.23   29            3    1056.18            12.3    1167.314400   
4        46.43   20            1    3902.00           464.3    4770.182750   

   line_cost  
0   80159.55  
1     331.79  
2     323.34  
3     732.95  
4    2627.70  

[5 rows x 22 columns]

In [14]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
df.head()



first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  line_cost  \
0            1  115164.00            75.0  125898.607500   80159.55   
1            0     523.95            14.9     588.693625     331.79   
2            6     527.28            16.8     594.407400     323.34   
3            3    1056.18            12.3    1167.314400     732.95   
4            1    3902.00           464.3    4770.182750    2627.70   

         profit  
0  45739.057500  
1    256.903625  
2    271.067400  
3    434.364400  
4   2142.482750  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above
order_ids_to_check = [2742071, 2173913, 6128929]
for order_id in order_ids_to_check:
    order_total = df[df['order_id'] == order_id]['total_price'].sum()
    print(f"Order ID {order_id} Total Price: {order_total}")


Order ID 2742071 Total Price: 139844.51202499997
Order ID 2173913 Total Price: 119838.35704999999
Order ID 6128929 Total Price: 752748.7454750001


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
spending_top_clients = df[df['client_id'].isin(top5_clients_list)].groupby('client_id')['total_price'].sum()
spending_top_clients



client_id
24741    7.667848e+07
33615    6.392918e+06
38378    9.172866e+06
46820    8.006333e+06
66037    8.746255e+06
Name: total_price, dtype: float64

In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
summary_df = df[df['client_id'].isin(top5_clients_list)].groupby('client_id').agg(
    total_units_purchased=('qty', 'sum'),
    total_shipping_price=('shipping_price', 'sum'),
    total_revenue=('total_price', 'sum'),
    total_profit=('profit', 'sum')
)
summary_df


total_units_purchased  total_shipping_price  total_revenue  \
client_id                                                               
24741                     239862                9365.6   7.667848e+07   
33615                      64313               12609.4   6.392918e+06   
38378                      73667               11895.0   9.172866e+06   
46820                      75768               11094.8   8.006333e+06   
66037                      43018               10017.3   8.746255e+06   

           total_profit  
client_id                
24741      3.610666e+07  
33615      2.033980e+06  
38378      2.955706e+06  
46820      2.589495e+06  
66037      3.126907e+06

In [18]:
print(df.columns.tolist())

['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id', 'order_date', 'order_week', 'order_year', 'item_id', 'category', 'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number', 'subtotal', 'shipping_price', 'total_price', 'line_cost', 'profit']


In [34]:
# Format the data and rename the columns to names suitable for presentation.
# Define a mapping for presentable column names
presentable_columns = {
    'client_id': 'Client ID',
    'order_id': 'Order ID',
    'order_date': 'Order Date',
    'order_week': 'Order Week',
    'order_year': 'Order Year',
    'item_id': 'Item ID',
    'category': 'Category',
    'subcategory': 'Subcategory',
    'unit_price': 'Unit Price',
    'unit_cost': 'Unit Cost',
    'unit_weight': 'Unit Weight',
    'qty': 'Quantity',
    'line_number': 'Line Number',
    'Shipping Price (Millions)': 'Shipping Cost (Millions)',
    'Total Price (Millions)': 'Total Cost (Millions)',
    'Line Cost (Millions)': 'Cost Per Line (Millions)',
    'Profit (Millions)': 'Profit Per Line (Millions)'
}
df.rename(columns=presentable_columns, inplace=True)

In [35]:
# Define the money columns. 
money_columns = ['shipping_price', 'total_price', 'line_cost', 'profit']
money_columns

['shipping_price', 'total_price', 'line_cost', 'profit']

In [36]:
# Define a function that converts a dollar amount to millions.
def to_millions(x):
    return x / 1e6

In [37]:
print("Money columns:", money_columns)
print("DataFrame columns:", df.columns)


Money columns: ['shipping_price', 'total_price', 'line_cost', 'profit']
DataFrame columns: Index(['first', 'last', 'job', 'phone', 'email', 'Client ID', 'Order ID',
       'Order Date', 'Order Week', 'Order Year', 'Item ID', 'Category',
       'Subcategory', 'Unit Price', 'Unit Cost', 'Unit Weight', 'Quantity',
       'Line Number', 'subtotal', 'shipping_price', 'total_price', 'line_cost',
       'profit'],
      dtype='object')


In [38]:
# Apply the currency_format_millions function to only the money columns. 
df[money_columns] = df[money_columns].fillna(0).apply(lambda x: x / 1e6)


In [39]:
# Rename the columns to reflect the change in the money format. 
money_columns = {
    'shipping_price': 'Shipping Price (Millions)',
    'total_price': 'Total Price (Millions)',
    'line_cost': 'Line Cost (Millions)',
    'profit': 'Profit (Millions)'
}
df.rename(columns=money_columns, inplace=True)

In [41]:
# Display the updated DataFrame to verify the changes
print(df)

           first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  Client ID  Order ID  Order Date  \
0       

In [43]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_df = df.sort_values(by='Profit (Millions)', ascending=False)
sorted_df.head()

first      last                                    job  \
21801    Julie  Anderson                    Exhibition designer   
15112   Dennis   Hawkins    Social research officer, government   
23929  Michele  Mcdonald  Emergency planning/management officer   
25972      Amy   Stewart              Psychologist, educational   
4063    Kendra   Garrett                 Pension scheme manager   

                       phone                            email  Client ID  \
21801             5933099462    anderson_better_4704@bell.biz      78965   
15112  001-214-888-5838x7410      hawkins.dennis.5195@lee.com      85931   
23929     435-931-4939x95137    mcdonaldm317@hess-jackson.com      38473   
25972     979.793.4884x57205           stewartamy@francis.com      86721   
4063            808.257.2029  garrettkendra8479@jefferson.com      24741   

       Order ID  Order Date  Order Week  Order Year  ... Unit Price Unit Cost  \
21801   3833568  2023-04-29          17        2023  ...     147.66    100.11   
15112   7263857  2023-03-05           9        2023  ...     323.42    183.84   
23929   4594154  2023-03-02           9        2023  ...     674.94    568.06   
25972   6057452  2023-03-21          12        2023  ...     100.40     73.05   
4063    5167984  2023-05-08          19        2023  ...     461.30    247.94   

      Unit Weight  Quantity  Line Number      subtotal  \
21801        0.09   3958244            1  5.844743e+08   
15112        2.21    637318            3  2.061214e+08   
23929        2.25    219645            5  1.482472e+08   
25972        9.22    977121            3  9.810295e+07   
4063         1.86    122166            7  5.635518e+07   

       Shipping Price (Millions)  Total Price (Millions)  \
21801               9.000000e-25            6.385382e-16   
15112               2.210000e-23            2.251876e-16   
23929               2.250000e-23            1.619601e-16   
25972               9.220000e-23            1.071776e-16   
4063                1.860000e-23            6.156805e-17   

       Line Cost (Millions)  Profit (Millions)  
21801          3.962598e-16       2.422784e-16  
15112          1.171646e-16       1.080231e-16  
23929          1.247716e-16       3.718853e-17  
25972          7.137878e-17       3.579879e-17  
4063           3.028986e-17       3.127819e-17  

[5 rows x 23 columns]